### ISSUE TITLE
Roadmap for Keras and TF2 Wrappers

### ISSUE THREAD

**Issue by seanpmorgan** [Trusted-AI/adversarial-robustness-toolbox#243]
[Keras 2.3 will be the last multi-backend Keras version](https://twitter.com/fchollet/status/1174018651449544704?s=20), and going forward it is recommended that users switch to `tf.keras`. Given this information, I think that it will be very confusing for end users to realize that a tf.keras model should be ran with `TensorFlowClassifierV2` instead of `KerasClassifier`. See https://github.com/IBM/adversarial-robustness-toolbox/issues/238

Ultimately this was done because [~97% of keras users are using TF](https://twitter.com/fchollet/status/1110354507395821568?s=20) and dropping Theano/CNTK makes it substantially easier to build features into the framework. I see a similar issue going on with ART and wonder if supporting this minority (that will only decrease) is worth all of the extra code/limitations?

**Comment by beat-buesser** [comment]
Hi @seanpmorgan Thank you for starting the discussion. Here are a few discussion points from our side:
- what would be the best way to maintain `KerasClassifier` to provide backward compatibility with `keras`?
- is it possible to distinguish between TensorFlow and `tf.keras` after the 2.0 release? `tf.keras` is now the recommended way of using TensorFlow, therefore having a classifier named `TensorFlowV2Classifier` which supports TF2.x models could make sense even if these models have been created with `tf.keras`. Could the confusion be solved with better documentation?

**Comment by beat-buesser** [comment]
We also could extend this discussion on a roadmap for `TensorFlowV2Classifier`. It already supports eager execution but we have not tested it with models/functions decorated with `@tf.function`. I think the current version might not support it. Supporting `@tf.function` seems to be important for [peak performance](https://www.tensorflow.org/tutorials/customization/performance).
@seanpmorgan Do you have any experience with `@tf.function` in TensorFlow 2.x

**Comment by seanpmorgan** [comment]
Hi @beat-buesser these are good points.

> what would be the best way to maintain KerasClassifier to provide backward compatibility with keras?

To me, it's more intuitive for KerasClassifier to refer to any keras model (multi-backend and tf.keras). There should be little to no breakage for using TF2 tf.keras models with the current wrapper because `model.compile()` needs to be ran prior to wrapping. The exception to this is if `model.compile(run_eagerly=True)` is used and then it could wrap with TFV2 Classifier instead. 

Eager execution / gradient tape is great for debugging/research but it will be slower than utilizing the compiled graph (caveat is if `@tf.function` is being properly utilized then its a graph representation anyway). By forcing TF2 `tf.keras` models which are graph compiled to be wrapped by the TFV2Classifier & GradientTape calls I suspect there would be significant performance hits compared to wrapping it in KerasClassifier which uses the Keras backend calls (need to test this).

> is it possible to distinguish between TensorFlow and tf.keras after the 2.0 release? tf.keras is now the recommended way of using TensorFlow, therefore having a classifier named TensorFlowV2Classifier which supports TF2.x models could make sense even if these models have been created with tf.keras. Could the confusion be solved with better documentation?

Yes, clear documentation/warnings that KerasClassifer refers to multi-backend keras would certainly help. 

> Do you have any experience with @tf.function in TensorFlow 2.x

Yes, a decent amount I will create an issue to test the current wrappers with models that utilize `@tf.function` in their calls.

**Comment by seanpmorgan** [comment]
So following the findings from #245, I would like to propose a refactoring of the TF/TF2/Keras wrapper paradigm. 

## Rationale for a change:
1) The reason we have multiple wrappers is to handle graph vs. eager execution. However, eager execution has been available since ~TF1.7. Naming the wrappers TF and TF2 is a mis-nomer.
2) As discussed in this thread, a single KerasWrapper would improve the user experience. Any model which uses KerasWrapper needs to be `model.compile()` first anyway so TF2 models can utilize the same wrapper. Shown [in this ART example notebook](https://github.com/IBM/adversarial-robustness-toolbox/blob/38e709f0261aa4945ca07892443b0bb49a361368/notebooks/art-for-tensorflow-v2-keras.ipynb). 
3) It would seem that a single TFWrapper with logic to utilize tf-eager or tf-graph should be sufficient. That way models that have user defined `train_steps` and `loss_objects` can be wrapped with TFWrapper, and models that subclass Keras can get all of the benefits of the Keras API.

Is this a solution that would be considered by the ART maintainers? If so I could write a more comprehensive proposal.

**Comment by beat-buesser** [comment]
@seanpmorgan Please apologise the delayed response and thank you for your proposal. Let me think about it for a moment and I'll get back to it very soon.
